[홍현진] Mission 4: 로그인 기능 구현

hjz1010 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- 로그인 기능 생성
		 
## 수정 사항들 자세한 내용
- LoginVeiw 작성
- 계정(이메일)과 비밀번호가 전달되었는지 확인 / 아닐 경우 KeyError 반환 
- 계정이 데이터베이스에 있는지 확인 후 없으면 INVALID_USER 반환
- 계정의 비밀번호와 입력한 비밀번호가 일치하는지 확인 후 틀리면 WRONG_PASSWORD 반환
- 모두 통과하면 SUCCESS 반환(로그인 성공)

- 앱 urls.py 에 http://127.0.0.1:8000/users/login 경로 추가 

## 기타 질문 및 특이 사항


## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

hjz1010 committed:
로그인 기능 구현 초안단계

hjz1010 committed:
LoginView urls 추가

hjz1010 committed:
오타(,) 수정

hjz1010 committed:
signup branch까지 병합된 main branch -> login branch로  병합

Merge branch 'main' into feature/hyeonjin-login

hjz1010 committed:
branch signup merge 후 충돌 발생한 부분 정리

hjz1010 committed:
필요없는 주석 삭제 및 align 정리

hjz1010 committed:
뭐가 바뀐건데

hjz1010 committed:
아니 뭐가 바뀐건데

j-eum commented on students/hyeonjin/users/views.py:
```diff
@@ -34,9 +34,29 @@ def post(self, request):
+            if usr.password == password:
+                return JsonResponse({"message":"SUCCESS"}, status=200)
+            else:
+                return JsonResponse({"message":"WRONG_PASSWORD"}, status=401)
```
```suggestion
                return JsonResponse({"message":"INVALID_USER"}, status=401)
```
이메일이 틀렸을때 비밀번호가 틀렸을때 구분하지 않고 둘다 같은 메세지를 보내는게 좋을것 같습니다
다르게 처리하면 보안상 문제가 있을 것 같아요!

hjz1010 replied:
오 보안 측면은 생각도 안하고 편의만 고려했어요 감사합니다!

j-eum reviewed (commented):
좋아요!

hjz1010 committed:
피어 리뷰 반영

- 아이디 에러와 비번 에러에 다른 에러메세지를 반환하면 보안상 위험할 수 있다는 지적
- 에러 메세지를 동일하게 수정 완료

hjz1010 commented on students/hyeonjin/users/views.py:
```diff
@@ -18,7 +18,7 @@ def post(self, request):
             others       = data['others']
             
-            email_validation = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
+            email_validation    = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
             password_validation = '^(?=.*[A-Za-z])(?=.*\d)(?=.*[$@$!%*#?&])[A-Za-z\d$@$!%*#?&]{8,}$'
```
정규표현식을 넣은 변수는 상수이므로 대문자로 변경해주세요.
또한 class 바깥에 선언해두면 다른 클래스에도 필요 시 활용할 수 있으니 위쪽으로 옮겨주세요.

hjz1010 replied:
정규표현식을 넣은 변수는 상수이므로 대문자로 변경해주세요(컨벤션 지키키!)
또한 class 바깥에 선언해두면 다른 클래스에도 필요 시 활용할 수 있으니 위쪽으로 옮겨주세요.
(동기분들 코드 보면서 배운 걸로 스스로 피어리뷰! ㅋㅋ)

hjz1010 committed:
피어 리뷰 반영2

- 상수는 대문자로
- 이메일과 비밀번호 정규표현식은 class 밖에 작성하여 다른 class에서도 활용할 수 있게

hjz1010 committed:
오타수정

hjz1010 committed:
Align 정리

Jon9Beom commented on students/hyeonjin/users/views.py:
```diff
@@ -34,9 +37,30 @@ def post(self, request):
+            email    = data['email']
+            password = data['password']
+
+            usr = User.objects.get(email=email)
+            if usr.password == password:
+                return JsonResponse({"message":"SUCCESS"}, status=200)
+            else:
+                return JsonResponse({"message":"INVALID_USER"}, status=401)
```
현진님 로직 잘 작성해주셨습니다!

다만 INPUT으로 받은 email과 password를 동시에 만족하는 User를 조회 하도록 코드를 작성하면 가독성이 더 좋아질 것 같습니다. (당연히 코드가 처리해야되는 로직도 간단해 지겠죠?)

이때, objects manager class의 `filter()` 메소드를 사용하는 것과 `exists()` 메소드를 사용하는 것의 차이가 무엇인지 설명해주세요.

추가로 filter method를 통해 구현한다면 57,58 line의 DoesNotExist 핸들링은 없어도 될 것 같습니다!
(filter는 get과 달리 만족하는 쿼리셋이 없다면 그저 빈 쿼리셋을 반환하기 때문입니다.)
```python
    data = json.loads(request.body)

    if not User.objects.filter(email = data["email"], password = data["password"]).exists():
		    return JsonResponse({'message' : 'INVALID_USER'}, status=401)

    return JsonResponse({'message' : 'LOGIN SUCCESS'}, status=200)
```

hjz1010 replied:
로그인 클래스에는 키에러는 확인하지 않아도 되나요?

hjz1010 replied:
```
class LoginView(View):

    def post(self, request):
        data = json.loads(request.body)

        try:
            email    = data['email']
            password = data['password']

            if User.objects.filter(email=email, password=password).exists():
                return JsonResponse({"message":"SUCCESS"}, status=200)
            return JsonResponse({"message":"INVALID_USER"}, status=401)    

        except KeyError:
            return JsonResponse({"message":"KEY_ERROR"}, status=400)
```

filter() 메소드를 사용하여 계정과  비밀번호가 동시에 일치하는 사용자를 데이터 베이스에서 찾고 
exists() 메소드를 사용하여 찾는 데이터가 존재하면 로그인 성공, 데이터가 없으면 에러를 반환한다. 
여기서 filter() 메소드는 get() 메소드와 달리 데이터가 존재하지 않을 경우 DoesNotExist 에러가 아니라 빈 쿼리셋을 반환하므로 
계정과 비밀번호의 예외처리를 따로 하지 않아도 된다. 
또한 빈 쿼리셋을 반환해도 if문에서 False로 처리되므로 exists()메소드 없이 로직을 작성해도 결과는 동일하지만
데이터의 존재만 확인하는 로직의 경우 exists() 메소드를 사용하는 것이 훨씬 가볍게 로직처리가 가능하다!

hjz1010 committed:
Login View 수정요청 반영

- email과 password를 동시에 만족하는 User를 조회하는 코드로 수정
-> 가독성 및 로직 개선
-> DoesNotExit 예외 따로 처리할 필요 X

Jon9Beom commented on students/hyeonjin/users/views.py:
```diff
@@ -34,9 +37,28 @@ def post(self, request):
+            email    = data['email']
+            password = data['password']
+
+            if User.objects.filter(email=email, password=password).exists():
+                return JsonResponse({"message":"SUCCESS"}, status=200)
+            return JsonResponse({"message":"INVALID_USER"}, status=401)    
```
현진님~ try 구문안에서 일어날 수 있는 예외 상황들은 분기로 처리하고 모든 것들이 정상적으로 통과가 될 때 성공을 반환하는게 좋습니다. 

코드블럭 통해 직접 예시로 설명드리겠습니다.

```python
class SignInView(View):
    def post(self, request):
        try:
            data = json.loads(request.body)

            email    = data['email']
            password = data['password']

            if not User.objects.filter(email=email, password=password).exists():
                return JsonResponse({'message':'INVALID_USER'}, status=401)

           if . . . 
                return JsonResponse ( . . . , status=400)
            
            return JsonResponse({'message':'SUCCESS'}, status=200)
        except KeyError:
            return JsonResponse({'message':'KEY_ERROR'}, status=400)
```

hjz1010 replied:
not 없이 코드를 작성해도 동일한 로직이 가능한데 이렇게 작성하는 이유가 궁금했었는데, 설명 감사합니다! :D

hjz1010 committed:
Login View 수정요청 반영2

- try 구문안에서 일어날 수 있는 예외 상황들은 분기로 처리하고
- 모든 것들이 정상적으로 통과가 될 때 성공을 반환하는게 좋다

Jon9Beom commented:
LGTM 🚀
